workers/udp.go

```diff
@@ -0,0 +1,130 @@
+package workers
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/frightenedmonkey/statsproxy/common"
+	"github.com/frightenedmonkey/statsproxy/config"
+	"net"
+	"strconv"
+	"strings"
+	"sync/atomic"
+)
+
+type Packet struct {
+	Bucket string
+	Raw    string
+}
+
+func StartUDPListeners(l *net.UDPConn) {
+	for i := 0; i < config.Service.Workers; i++ {
+		go udpListener(l)
+	}
+}
+
+func StartUDPWriters() {
+	for i := 0; i < config.Service.WriterMultiplier; i++ {
+		for _, host := range config.Service.Statsd.Hosts {
+			go hostWriter(host)
+		}
+	}
+}
+
+func hostWriter(host config.Host) {
+	coalescedPacket := make([]byte, 0)
+	for {
+		newMessage := <-common.HostChans[host]
+		if len(coalescedPacket)+len([]byte(newMessage)) > 1400 {
+			common.HostConns[host].Write(coalescedPacket)
+			coalescedPacket = []byte(newMessage)
+		} else {
+			coalescedPacket = append(coalescedPacket, []byte(newMessage)...)
+		}
+	}
+}
+
+func udpListener(l *net.UDPConn) {
+	message := make([]byte, config.Service.MaxUDPPacketSize)
+	for {
+		n, remaddr, err := l.ReadFromUDP(message)
+		if err != nil {
+			common.Logger.Err(fmt.Sprintf("ERROR: reading UDP packet from %+v - %s",
+				remaddr, err))
+			continue
+		}
+
+		dataHandler(message[:n])
+	}
+}
+
+func dataHandler(data []byte) {
+	packets := parseMessage(data)
+	for _, packet := range packets {
+		packetHandler(packet)
+	}
+}
+
+func parseMessage(data []byte) []Packet {
+	var input []byte
+	var packets = make([]Packet, 0)
+
+	for _, line := range bytes.Split(data, []byte("\n")) {
+		if len(line) == 0 {
+			continue
+		}
+
+		input = line
+
+		index := bytes.IndexByte(input, ':')
+		if index < 0 || index == len(input)-1 {
+			common.Logger.Err(fmt.Sprintf("ERROR: failed to parse line: %s\n",
+				string(line)))
+			continue
+		}
+
+		name := input[:index]
+
+		packet := Packet{
+			Bucket: string(name),
+			Raw:    string(line) + "\n",
```

Comment:
As far as I'm aware, the coalesced metrics sent in are separated via "\n", and the logic for how the messages are split up removes the newline at the end, so it makes sense to add it back for line termination purposes. In other words, I don't believe we can get a newline at the end by this point, but even if there were, I'm pretty sure it wouldn't be problematic.